### PR TITLE
refactor: migrate from `enquirer` to `prompts`

### PR DESCRIPTION
**Pull request by kanadgupta**
| 🚥 Fix RM-5116 |
| :-- |

## 🧰 Changes

Migrating our prompts system from the universally-disliked [`enquirer`](https://github.com/enquirer/enquirer) to [`prompts`](https://github.com/terkelg/prompts). In addition to it being a far better from a DX and testing standpoint, we also get a few nice helpful cues from `prompts` when a user is required to take an action (note the grayed out "use arrow-keys" guidance):

image

Overall, the prompt object patterns are fairly similar and there shouldn't be a whole lot of material changes for users (except in maybe the `login` command since that was previously using the ancient [`read`](https://npm.im/read) package).

### Which Commands Are Affected?

All areas of `rdme` that use a prompts-based system have been migrated over, namely:

- [x] Version selection (prompt: selecting project versions in the `categories`, `docs`, `openapi`, and `versions` command families)
- [x] `openapi` (prompt: selecting a spec)
- [x] `versions:create` (prompt: setting properties on the version you want to create)
- [x] `versions:update` (prompt: similar to above[^2])
- [x] `login` (prompt: logging in!)

### Other Small Cleanups 🧹 

Because this work got me looking into commands that don't get a lot of love, here are a bunch of other miscellaneous changes in this PR:
 - [x] I cleaned a bunch of types and added a bunch of tests in the `versions` group of commands (and honestly there's still a lot of work left to be done) and may or may not have fixed a few random bugs in the process. Not sure how much these commands are being used, I'd honestly love if we did away with them at some point 👀 
 - [x] Swapped out [`isemail`](https://npm.im/isemail) for [`vaildator`](https://npm.im/vaildator) since it's better maintained and a bit smaller of a dep.
 - [x] Shortened our debug command from `debug:bin` to just `debug`.
 - [x] Finally got our `openapi` command test coverage to 💯 
 - [x] Added a smol send off when someone exits the prompt system early:

image

## 🧬 QA & Testing

Tests (particularly for the version commands) should be far more resilient, so if those pass then that's a really good sign[^1]. I've tried a bunch of different user journeys locally (and backfilled test coverage where I easily could) and everything appears to work well!

[^1]: I still **cannot** get over how almost all of our `enquirer`-mocked tests were simply not testing anything at all for like... 3+ years now. Good riddance 🖕🏽 
[^2]: you'd think it was similar enough to `create` that this would be easy... but they're annoyingly different

### CODE REVIEW

File: __tests__/lib/prompts.test.ts [hunk 1]
```diff
@@ -36,58 +38,34 @@ const getSpecs = () => {
...
-            // eslint-disable-next-line jest/no-conditional-expect
-            await expect(prompt.skip()).resolves.toBe(true);
-            break;
-
-          case 'newVersion':
-            // eslint-disable-next-line no-param-reassign
-            prompt.value = '1.2.1';
-            await prompt.submit();
-        }
-      });
```
Comment (erunion) [inline __tests__/lib/prompts.test.ts:65]: giphy

File: __tests__/lib/prompts.test.ts [hunk 2]
```diff
@@ -101,19 +79,11 @@ describe('prompt test bed', () => {
...
       );
 
-      expect(answer.option).toBe('create');
+      expect(answer).toStrictEqual({ option: 'create' });
     });
 
     it('should return specId if user chooses to update file', async () => {
-      jest.mock('enquirer');
-      enquirer.on('prompt', async prompt => {
-        await prompt.keypress(null, { name: 'down' });
```
Comment (erunion) [inline __tests__/lib/prompts.test.ts:110]: We love to be forced to test a CLI UX like a React component don't we folks.

File: src/cmds/login.ts [hunk 3]
```diff
@@ -1,22 +1,16 @@
 import type { CommandOptions } from '../lib/baseCommand';
 
-import { promisify } from 'util';
-
 import chalk from 'chalk';
 import config from 'config';
-import { validate as isEmail } from 'isemail';
-import readPkg from 'read';
```
Comment (erunion) [inline src/cmds/login.ts:8]: Very excited to only have one user input library to deal with now.

File: src/cmds/login.ts [hunk 4]
```diff
@@ -50,23 +44,39 @@ export default class LoginCommand extends Command {
...
+        type: 'text',
+        name: 'email',
+        message: 'What is your email address?',
+        initial: configStore.get('email'),
+        validate(val) {
+          return isEmail(val) ? true : 'Please provide a valid email address.';
+        },
+      },
+      {
+        type: 'invisible',
```
Comment (erunion) [inline src/cmds/login.ts:60]: :neat-rgb:

File: src/cmds/openapi.ts [hunk 5]
```diff
@@ -227,12 +228,11 @@ export default class OpenAPICommand extends Command {
       Command.debug(`api settings list response payload: ${JSON.stringify(apiSettingsBody)}`);
       if (!apiSettingsBody.length) return createSpec();
 
-      const { option }: { option: 'create' | 'update' } = await prompt(
+      const { option }: Answers<string> = await promptTerminal(
         promptHandler.createOasPrompt(apiSettingsBody, parsedDocs, totalPages, getSpecs)
       );
-
       Command.debug(`selection result: ${option}`);
-      if (!option) return null;
```
Comment (erunion) [inline src/cmds/openapi.ts:235]: Is `option` never going to be `null` now?

File: src/cmds/openapi.ts [hunk 6]
```diff
@@ -227,12 +228,11 @@ export default class OpenAPICommand extends Command {
       Command.debug(`api settings list response payload: ${JSON.stringify(apiSettingsBody)}`);
       if (!apiSettingsBody.length) return createSpec();
 
-      const { option }: { option: 'create' | 'update' } = await prompt(
+      const { option }: Answers<string> = await promptTerminal(
```
Comment (erunion) [inline src/cmds/openapi.ts:231]: I know we can't have `{ option: 'create' | 'update' }` typing anymore but the `Answers<string>` typing here is a little weird because it doesn't seem to be doing anything. Maybe remove it until we figure out how to better work with this `Answers<>` generic and our destructured answers?

Screen Shot 2022-08-18 at 6 12 49 PM

```suggestion
      const { option } = await promptTerminal(
```

**Review by erunion — APPROVED** [review]
The most exciting PR in this repository in a long time.

File: src/lib/prompts.ts [hunk 7]
```diff
@@ -24,56 +28,61 @@ type ParsedDocs = {
...
     },
     {
-      type: 'select',
+      type: (prev, values) => {
+        return (selectOnly ? false : values.option !== 'update') ? null : 'select';
+      },
       name: 'versionSelection',
       message: 'Select your desired version',
-      skip() {
-        return selectOnly ? false : this.enquirer.answers.option !== 'update';
```
[the remark this answers is not in the document]
Comment (kanadgupta) [reply src/lib/prompts.ts:46]: too bad it's arguably more confusing in the `type` field lol 🤡

File: src/lib/prompts.ts [hunk 8]
```diff
@@ -84,19 +93,19 @@ const updateOasPrompt = (
...
     choices: specOptions(specList, parsedDocs, currPage, totalPages),
-    async result(spec: string) {
+    async format(spec: string) {
       if (spec === 'prev') {
         try {
           const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
           const newParsedDocs = parse(newSpecs.headers.get('link'));
           const newSpecList = await newSpecs.json();
-          const { specId }: { specId: string } = await prompt(
+          const { specId }: Answers<string> = await promptTerminal(
```
[the remark this answers is not in the document]
Comment (kanadgupta) [reply src/lib/prompts.ts:108]: how i've been learning TS so far: blindly trying intellisense suggestions until the squiggly lines go away

File: src/lib/prompts.ts [hunk 9]
```diff
@@ -157,75 +162,56 @@ export function createOasPrompt(
...
   isUpdate?: {
-    is_stable: string;
+    is_stable: boolean;
   }
-) {
+): PromptObject[] {
   return [
     {
-      type: 'select',
+      type: opts.fork || isUpdate ? null : 'select',
```
[the remark this answers is not in the document]
Comment (kanadgupta) [reply src/lib/prompts.ts:172]: yeah honestly this prompt flow was extremely hard to follow when rewriting this... I still think there a ton of opportunities to clean up our logic for these commands (akin to what you worked on in https://github.com/readmeio/rdme/pull/449) but these commands are likely not used a whole lot so I decided it wasn't worth the time 🤷🏽‍♂️

File: src/lib/prompts.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (erunion) [reply src/lib/prompts.ts:108]: just like the rest of us

File: __tests__/lib/prompts.test.ts (hunk 2 above)
Comment (kanadgupta) [reply __tests__/lib/prompts.test.ts:110]: such a goddamn mess

File: src/cmds/openapi.ts (hunk 6 above)
Comment (kanadgupta) [reply src/cmds/openapi.ts:231]: will do! i'm with you, it's not super helpful in its current form. i was kinda struggling with how to mangle with this generic and make it... less generic lol.

Comment (kanadgupta) [reply src/cmds/openapi.ts:231]: i'm gonna rip them out from everywhere!

File: __tests__/cmds/openapi.test.ts [hunk 10]
```diff
@@ -429,18 +489,15 @@ describe('rdme openapi', () => {
...
+      prompts.inject(['create', '1.0.1']);
 
       const registryUUID = getRandomRegistryId();
 
       const mock = getAPIMock()
         .get('/api/v1/version')
         .basicAuth({ user: key })
         .reply(200, [{ version: '1.0.0' }])
-        .post('/api/v1/version')
+        .post('/api/v1/version', { from: '1.0.0', version: '1.0.1', is_stable: false })
```
[the remark this answers is not in the document]
Comment (kanadgupta) [reply __tests__/cmds/openapi.test.ts:500]: yeaaaah a strict mode would be super helpful, i'm sure there are tons of gaping holes in this test bed

File: src/cmds/openapi.ts (hunk 5 above)
Comment (kanadgupta) [reply src/cmds/openapi.ts:235]: Yeah I couldn't figure out how to reproduce that when writing tests... if anything we should be throwing an error there rather than returning `null`. I imagine it's coming from our pagination stuff, which I added a little test TODO for here:
https://github.com/readmeio/rdme/blob/e5845e26a9f223c19f92eafe2299693d602589dc/__tests__/cmds/openapi.test.ts#L352

**Review by erunion — APPROVED** [review]
LFG
